src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,193 @@
+            };
+            BuildManager.DefaultBuildManager.BeginBuild(parameters);
+
+            // Do a restore first (equivalent to MSBuild's "implicit restore", i.e., `/restore`).
+            // See https://github.com/dotnet/msbuild/blob/a1c2e7402ef0abe36bf493e395b04dd2cb1b3540/src/MSBuild/XMake.cs#L1838.
+            var restoreRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection, addGlobalProperties: static (globalProperties) =>
+                {
+                    globalProperties["MSBuildRestoreSessionId"] = Guid.NewGuid().ToString("D");
+                    globalProperties["MSBuildIsRestoring"] = bool.TrueString;
+                }),
+                targetsToBuild: ["Restore"],
+                hostServices: null,
+                BuildRequestDataFlags.ClearCachesAfterBuild | BuildRequestDataFlags.SkipNonexistentTargets | BuildRequestDataFlags.IgnoreMissingEmptyAndInvalidImports | BuildRequestDataFlags.FailOnUnresolvedSdk);
+            var restoreResult = BuildManager.DefaultBuildManager.BuildRequest(restoreRequest);
+            if (restoreResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            // Then do a build.
+            var buildRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection),
+                targetsToBuild: ["Build"]);
+            var buildResult = BuildManager.DefaultBuildManager.BuildRequest(buildRequest);
+            if (buildResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            BuildManager.DefaultBuildManager.EndBuild();
+            return 0;
+        }
+        catch (Exception e)
+        {
+            Console.Error.WriteLine(e.Message);
+            return 1;
+        }
+        finally
+        {
+            foreach (var (key, value) in savedEnvironmentVariables)
+            {
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            binaryLogger?.Shutdown();
+            consoleLogger.Shutdown();
+        }
+
+        static ILogger? GetBinaryLogger(string[] args)
+        {
+            for (int i = args.Length - 1; i >= 0; i--)
+            {
+                var arg = args[i];
+                if (arg.StartsWith("/bl:") || arg.Equals("/bl")
+                    || arg.StartsWith("--binaryLogger:") || arg.Equals("--binaryLogger")
+                    || arg.StartsWith("-bl:") || arg.Equals("-bl"))
+                {
+                    return new BinaryLogger
+                    {
+                        Parameters = arg.IndexOf(':') is >= 0 and var index
+                            ? arg[(index + 1)..]
+                            : "msbuild.binlog",
+                    };
+                }
+            }
+
+            return null;
+        }
+    }
+
+    public ProjectInstance CreateProjectInstance(ProjectCollection projectCollection)
```

Comment:
Is there a reason this overload is needed? The below one with the optional should work for the call sites I see.

Reply:
Yes, there is a delegate created from this one here:

https://github.com/dotnet/sdk/blob/8094d4938812d7ddf2a6b971d70b747fbeb3a1e6/src/Cli/dotnet/commands/dotnet-run/RunCommand.cs#L262

The one with the optional argument wouldn't work (would need a lambda wrapper which seems less efficient).

Also this one is public whereas the one with the optional arg isn't exposed, it's private.